ENIGMAsystem/SHELL/Widget_Systems/Native/libdlgmod.cpp

```diff
@@ -0,0 +1,233 @@
+    libdlgmod_get_string = external_define(LIBDLGMOD, "get_string", 0, ty_string, 2, ty_string, ty_string);
+    libdlgmod_get_password = external_define(LIBDLGMOD, "get_password", 0, ty_string, 2, ty_string, ty_string);
+    libdlgmod_get_integer = external_define(LIBDLGMOD, "get_integer", 0, ty_real, 2, ty_string, ty_real);
+    libdlgmod_get_passcode = external_define(LIBDLGMOD, "get_passcode", 0, ty_real, 2, ty_string, ty_real);
+    libdlgmod_get_open_filename = external_define(LIBDLGMOD, "get_open_filename", 0, ty_string, 2, ty_string, ty_string);
+    libdlgmod_get_open_filename_ext = external_define(LIBDLGMOD, "get_open_filename_ext", 0, ty_string, 4, ty_string, ty_string, ty_string, ty_string);
+    libdlgmod_get_open_filenames = external_define(LIBDLGMOD, "get_open_filenames", 0, ty_string, 2, ty_string, ty_string);
+    libdlgmod_get_open_filenames_ext = external_define(LIBDLGMOD, "get_open_filenames_ext", 0, ty_string, 4, ty_string, ty_string, ty_string, ty_string);
+    libdlgmod_get_save_filename = external_define(LIBDLGMOD, "get_save_filename", 0, ty_string, 2, ty_string, ty_string);
+    libdlgmod_get_save_filename_ext = external_define(LIBDLGMOD, "get_save_filename_ext", 0, ty_string, 4, ty_string, ty_string, ty_string, ty_string);
+    libdlgmod_get_directory = external_define(LIBDLGMOD, "get_directory", 0, ty_string, 1, ty_string);
+    libdlgmod_get_directory_alt = external_define(LIBDLGMOD, "get_directory_alt", 0, ty_string, 2, ty_string, ty_string);
+    libdlgmod_get_color = external_define(LIBDLGMOD, "get_color", 0, ty_real, 1, ty_real);
+    libdlgmod_get_color_ext = external_define(LIBDLGMOD, "get_color_ext", 0, ty_real, 2, ty_real, ty_string);
+    libdlgmod_widget_get_caption = external_define(LIBDLGMOD, "widget_get_caption", 0, ty_string, 0);
+    libdlgmod_widget_set_caption = external_define(LIBDLGMOD, "widget_set_caption", 0, ty_real, 1, ty_string);
+    libdlgmod_widget_get_owner = external_define(LIBDLGMOD, "widget_get_owner", 0, ty_string, 0);
+    libdlgmod_widget_set_owner = external_define(LIBDLGMOD, "widget_set_owner", 0, ty_real, 1, ty_string);
+    libdlgmod_widget_get_icon = external_define(LIBDLGMOD, "widget_get_icon", 0, ty_string, 0);
+    libdlgmod_widget_set_icon = external_define(LIBDLGMOD, "widget_set_icon", 0, ty_real, 1, ty_string);
+    libdlgmod_widget_get_system = external_define(LIBDLGMOD, "widget_get_system", 0, ty_string, 0);
+    libdlgmod_widget_set_system = external_define(LIBDLGMOD, "widget_set_system", 0, ty_real, 1, ty_string);
+    libdlgmod_widget_get_button_name = external_define(LIBDLGMOD, "widget_get_button_name", 0, ty_string, 1, ty_real);
+    libdlgmod_widget_set_button_name = external_define(LIBDLGMOD, "widget_set_button_name", 0, ty_real, 2, ty_real, ty_string);
+    libdlgmod_widget_get_canceled = external_define(LIBDLGMOD, "widget_get_canceled", 0, ty_real, 0);
+  }
+
+  int show_message(std::string str) {
+    return (int)external_call(libdlgmod_show_message, str);

```

Comment:
[Discussed offline] The use of FFI routines adds some overhead to the calling convention, but for a dialog system this should be negligible...